editors/code/src/installation/extension.ts

```diff
@@ -0,0 +1,144 @@
+
+    if (currentExtReleaseDate === null) {
+        void vscode.window.showErrorMessage(
+            "Nightly release date must've been set during the installation. " +
+            "Did you download and install the nightly .vsix package manually?"
+        );
+        throw new Error("Nightly release date was not set in globalStorage");
+    }
+
+    const dateNow = new Date;
+    const hoursSinceLastUpdate = diffInHours(currentExtReleaseDate, dateNow);
+    log.debug(
+        "Current rust-analyzer nightly was downloaded", hoursSinceLastUpdate,
+        "hours ago, namely:", currentExtReleaseDate, "and now is", dateNow
+    );
+
+    if (hoursSinceLastUpdate < HEURISTIC_NIGHTLY_RELEASE_PERIOD_IN_HOURS) {
+        return;
+    }
+    if (!await askToDownloadProperExtensionVersion(config, "The installed nightly version is most likely outdated. ")) {
+        return;
+    }
+
+    await tryDownloadNightlyExtension(config, releaseInfo => {
+        assert(
+            currentExtReleaseDate.getTime() === config.installedNightlyExtensionReleaseDate.get()?.getTime(),
+            "Other active VSCode instance has reinstalled the extension"
+        );
+
+        if (releaseInfo.releaseDate.getTime() === currentExtReleaseDate.getTime()) {
+            vscode.window.showInformationMessage(
+                "Whoops, it appears that your nightly version is up-to-date. " +
+                "There might be some problems with the upcomming nightly release " +
+                "or you traveled too far into the future. Sorry for that 😅! "
+            );
+            return false;
+        }
+        return true;
+    });
+}
+
+async function askToDownloadProperExtensionVersion(config: Config, reason = "") {
+    if (!config.askBeforeDownload) return true;
+
+    const stableOrNightly = config.updatesChannel === UpdatesChannel.Stable ? "stable" : "latest nightly";
+
+    // In case of reentering this function and showing the same info message
+    // (e.g. after we had shown this message, the user changed the config)
+    // vscode will dismiss the already shown one (i.e. return undefined).
+    // This behaviour is what we want, but likely it is not documented
+
+    const userResponse = await vscode.window.showInformationMessage(
+        reason + `Do you want to download the ${stableOrNightly} rust-analyzer extension ` +
+        `version and reload the window now?`,
+        "Download now", "Cancel"
+    );
+    return userResponse === "Download now";
+}
+
+/**
+ * Shutdowns the process in case of success (i.e. reloads the window) or throws an error.
+ *
+ * ACHTUNG!: this function has a crazy amount of state transitions, handling errors during
+ * each of them would result in a ton of code (especially accounting for cross-process
+ * shared mutable `globalState` access). Enforcing no reentrancy for this is best-effort.
+ */
+const tryDownloadNightlyExtension = notReentrant(async function tryDownloadNightlyExtension(
```

Comment:
```suggestion
const tryDownloadNightlyExtension = notReentrant(async ()
```

That way, direct recursive calls will also go via wrapper.

Reply:
I intentionally used a named lambda here, because, we log `fn.name` when reentrancy invariant is broken. Or do you think this won't be necessary as we could follow the stacktrace?

Reply:
"decorating" named function in JS is a footgun, as recursive calls (or nested closures) will use undercoated name. This problems seems worse than slight disadvantage in reading the stack trace.

But no strong opinion here. TBH, I would just try very hard to make all control flow here blindly obvious, such that these checks are not required.

Reply:
Oh, I just understood what you meant by *"direct recursive calls"* and you do have a point!

Okay, let's not rush and strive to make the control flow more evident.
Do you think it is sufficient to handle reentrancy closer to `ensure*()` functions' call site? 

From my perspective, it would be better to explicitly add cancellation support for `ensureProperExtensionVersion()` and not rely on undocumented vscode behavior as [stated here](https://github.com/rust-analyzer/rust-analyzer/blob/9789f984cbe20a7fca1354e542be7bb345d795eb/editors/code/src/installation/extension.ts#L92-L95).

Besides that I guess there is no need for `downloadServer()` cancellation support right now as we call this function only once, so `notReentrant()` for it is really a debug assertion, whilst in case of `tryDownloadNightlyExtension()` it do can blow at runtime and we handle this by throwing `AssertionError` in my current PR :sweat_smile:, thus my cancellation support proposal.

Reply:
I think the current solution is fine. I do believe that the best way to arrive at the good code is to actually rush something working in production, and then start incrementally improving it. I'd rather get this in some form in today's release, the behavior shouldn't be significantly different unless the user explicitly requests nightly. 

The code in question is not in my cache right now, so I can't give specific advice, but I'll look closely at it later today :)

Reply:
Okay, thanks, I'll change this to arrow functions and merge for now then!